lib/web/apiserver.go

```diff
@@ -938,33 +926,40 @@ func (h *Handler) githubCallback(w http.ResponseWriter, r *http.Request, p httpr
 
 	response, err := h.cfg.ProxyClient.ValidateGithubAuthCallback(r.URL.Query())
 	if err != nil {
-		logger.Warnf("Error while processing callback: %v.", err)
-		// redirect to an error page
-		pathToError := url.URL{
-			Path: "/web/msg/error/login_failed",
-			RawQuery: url.Values{"details": []string{
-				"Unable to process callback from Github."}}.Encode(),
-		}
-		http.Redirect(w, r, pathToError.String(), http.StatusFound)
+		logger.WithError(err).Error("Error while processing callback.")
+		http.Redirect(w, r, loginFailedBadCallbackURL, http.StatusFound)
 		return nil, nil
 	}
+
 	// if we created web session, set session cookie and redirect to original url
 	if response.Req.CreateWebSession {
-		err = csrf.VerifyToken(response.Req.CSRFToken, r)
-		if err != nil {
-			logger.Warnf("Unable to verify CSRF token: %v.", err)
-			return nil, trace.AccessDenied("access denied")
+		logger.Infof("Redirecting to web browser.")
+
+		if err := csrf.VerifyToken(response.Req.CSRFToken, r); err != nil {
+			logger.WithError(err).Error("Unable to verify CSRF token.")
+			http.Redirect(w, r, loginFailedURL, http.StatusFound)
+			return nil, nil
 		}
-		logger.Infof("Callback is redirecting to web browser.")
-		err = SetSessionCookie(w, response.Username, response.Session.GetName())
-		if err != nil {
-			return nil, trace.Wrap(err)
+
+		if err := SetSessionCookie(w, response.Username, response.Session.GetName()); err != nil {
+			logger.WithError(err).Error("Unable to set session cookie.")
```

Comment:
So this is not a huge deal, but I'm wondering if you have considered making a some sort of wrapper method for these handlers which would do this logging/redirect instead of adding the same 2 lines of code to every "if err != nil" block. Would make the code a bit cleaner.

Reply:
revised

It made sense to create a redirect handler wrapper b/c we didn't use the return type values of the default handler. But I kept the logging of errors closer to origin b/c it was easier to customize